feat: Implement barebones Server Subscriptions support.

FayeDel opened the pull request:
## About

This PR title essentially explains itself.

Server Subscriptions, formerly Premium Membership, have been released widespread towards all community-based servers that resides in the US, hence the ready-to-review PR.

This correlates to an earlier PR: #630 

## Checklist

- [X] I've ran `pre-commit` to format and lint the change(s) made.
- [X] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [X] I've made this pull request for/as: (check all that apply)
  - [X] Documentation
  - [ ] Breaking change
  - [X] New feature/enhancement
  - [ ] Bugfix

i0bs commented:
Premium memberships now offer information on multiple subscriptions present on a guild member. See [here for more.](https://support.discord.com/hc/en-us/articles/4415163187607-Premium-Memberships-for-Servers)

FayeDel committed:
Merge pull request #1121 from interactions-py/unstable

chore: Merge unstable to stable

FayeDel commented:
Marking this ready to review *only* because the feature's live;
(This PR still needs to be updated to latest unstable however)

i0bs commented on interactions/api/models/role.py:
```diff
@@ -18,13 +18,27 @@ class RoleTags(DictSerializerMixin):
-
-    # TODO: Figure out what actual type it returns, all it says is null.
+    subscription_listing_id: Optional[Snowflake] = field(converter=Snowflake, default=None)
+    purchasable_or_has_subscribers: Optional[Any] = field(default=None)
```
Appears to return `bool`

FayeDel committed:
docs, feat: Implement barebones Premium Membership support.

(Recommitted via very old branch, origin commit hash 936a0f8c85134a93fe190c496eceb267ca6b4633)

FayeDel committed:
docs, feat: Implement unofficial new MessageType and AuditLogEvent type.

FayeDel committed:
Merge branch 'unstable' into unstable-undoc-membership

Damego commented on interactions/api/models/role.py:
```diff
@@ -19,16 +19,19 @@ class RoleTags(DictSerializerMixin):
-    :ivar Optional[Snowflake] bot_id: The id of the bot this role belongs to
-    :ivar Optional[Snowflake] integration_id: The id of the integration this role belongs to
-    :ivar Optional[Any] premium_subscriber: Whether if this is the guild's premium subscriber role
+    :ivar Optional[Snowflake] bot_id?: The id of the bot this role belongs to
+    :ivar Optional[Snowflake] integration_id?: The id of the integration this role belongs to
+    :ivar Optional[Any] premium_subscriber?: Whether if this is the guild's server subscription role
+    :ivar Optional[Snowflake] subscription_listing_id?: The id of the linked server subscription role, if any.
+    :ivar Optional[bool] purchasable_or_has_subscribers?: Whether the role can be purchasable via server subscription or that the role contains subscribers, if any.
```
We refused from `?` in docs

FayeDel committed:
docs: Remove `?` on RoleTags object attribute docs.

FayeDel committed:
Merge remote-tracking branch 'origin/unstable-undoc-membership' into unstable-undoc-membership

Damego commented on interactions/api/models/gw.py:
```diff
@@ -258,10 +258,14 @@ class GuildJoinRequest(DictSerializerMixin):
 
     user_id: Snowflake = field(converter=Snowflake)
     guild_id: Snowflake = field(converter=Snowflake)
+    request: Optional[Any] = field(default=None)
```
There are no info about `request` to properly annotate it?

FayeDel replied:
The only thing I've gotten from my last testing (months ago) was a dictionary, however I don't have the logs for it nor is it documented in the docs (nor can I test it further at the moment) nor if I know if it's still there now compared to local testing.

Damego replied:
lol. Okay then

FayeDel committed:
docs, feat: Include server-subscriptions related error codes, enums, and proper nomenclature.

(according to discord-api-docs, ebebc249cbe817c0927b17189b533f4d60715961)

Damego reviewed (approved):
Looks good to me
